StRoot/StGeant4Maker/StGeant4Maker.cxx

```diff
@@ -0,0 +1,1539 @@
+      g2t_hit.x[2]      = (hit->position_in[2] + hit->position_out[2]) * 0.5;
+      g2t_hit.tof       = (hit->position_in[3] + hit->position_out[3]) * 0.5;
+      g2t_hit.p[0]      = (hit->momentum_in[0] + hit->momentum_out[0]) * 0.5;
+      g2t_hit.p[1]      = (hit->momentum_in[1] + hit->momentum_out[1]) * 0.5;
+      g2t_hit.p[2]      = (hit->momentum_in[2] + hit->momentum_out[2]) * 0.5;
+      g2t_hit.s_track   = hit->length;
+      
+      table -> AddAt( &g2t_hit );     
+
+      int idtruth = hit->idtruth;
+      g2t_track_st* trk = (g2t_track_st*)track->At(idtruth-1);
+
+      trk->n_vpd_hit++;
+      
+    }
+  } 
+} sd2table_vpd; 
+struct SD2Table_MTD {
+  void operator()( StSensitiveDetector* sd, St_g2t_mtd_hit* table, St_g2t_track* track ) {
+    
+    TString sdname = sd->GetName();
+
+    // Retrieve the hit collection 
+    StTrackerHitCollection* collection = (StTrackerHitCollection *)sd->hits();
+    // Iterate over all hits
+    for ( auto hit : collection->hits() ) {
+
+      g2t_mtd_hit_st g2t_hit; memset(&g2t_hit,0,sizeof(g2t_mtd_hit_st)); 
+      
+      g2t_hit.id        = hit->id;
+      // TODO: add pointer to next hit on the track 
+      g2t_hit.track_p   = hit->idtruth;
+      g2t_hit.volume_id = hit->volId;
+      g2t_hit.de        = hit->de;
+      g2t_hit.ds        = hit->ds;
+
+      g2t_hit.xglobal[0]      = (hit->position_in[0] + hit->position_out[0]) * 0.5;
+      g2t_hit.xglobal[1]      = (hit->position_in[1] + hit->position_out[1]) * 0.5;
+      g2t_hit.xglobal[2]      = (hit->position_in[2] + hit->position_out[2]) * 0.5;
+
+      g2t_hit.tof       = (hit->position_in[3] + hit->position_out[3]) * 0.5;
+      g2t_hit.p[0]      = (hit->momentum_in[0] + hit->momentum_out[0]) * 0.5;
+      g2t_hit.p[1]      = (hit->momentum_in[1] + hit->momentum_out[1]) * 0.5;
+      g2t_hit.p[2]      = (hit->momentum_in[2] + hit->momentum_out[2]) * 0.5;
+      g2t_hit.s_track   = hit->length;
+      
+      table -> AddAt( &g2t_hit );     
+
+      int idtruth = hit->idtruth;
+      g2t_track_st* trk = (g2t_track_st*)track->At(idtruth-1);
+
+      trk->n_mtd_hit++;
+      
+    }
+  } 
+} sd2table_mtd; 
+
+//________________________________________________________________________________________________
+TGeant4* gG4 = 0;
+TGeant3TGeo* gG3 = 0;
+
+// Function to process one event
+std::function<void(void)> trigger;
+
+//________________________________________________________________________________________________
+// Pointer to the maker so we can forward VMC calls there
+static StGeant4Maker* _g4maker = 0;
+//________________________________________________________________________________________________
+StarParticleData &particleData = StarParticleData::instance();
+//________________________________________________________________________________________________
+StarVMCApplication::StarVMCApplication( const Char_t *name, const Char_t *title, double zmax, double rmax, std::string multi, StMCParticleStack* stack ) : 
```

Comment:
You will note that there are two class implementations in this file.  StGeant4Maker and StarVMCApplication.  The VMC application creates and steers the simulation, while the StGeant4Maker serves as the interface and bridge between our framework and the application.